flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamIterationHead.java

```diff
@@ -119,13 +183,154 @@ protected void cleanup() throws Exception {
+	 */
+	public static class UpstreamLogger<IN> extends AbstractStreamOperator<IN> implements OneInputStreamOperator<IN, IN> {
+
+		private final StreamConfig config;
+
+		private LinkedList<ListState<StreamRecord<IN>>> slicedLog = new LinkedList<>();
+
+		private UpstreamLogger(StreamConfig config) {
+			this.config = config;
+		}
+
+		public void logRecord(StreamRecord<IN> record) throws Exception {
+			if (!slicedLog.isEmpty()) {
+				slicedLog.getLast().add(record);
+			}
+		}
+
+		public void createSlice(String sliceID) throws Exception {
+			ListState<StreamRecord<IN>> nextSlice =
+				getOperatorStateBackend().getOperatorState(new ListStateDescriptor<>(sliceID,
+					config.<StreamRecord<IN>>getTypeSerializerOut(getUserCodeClassloader())));
+			slicedLog.addLast(nextSlice);
+		}
+
+		public void discardSlice() {
+			ListState<StreamRecord<IN>> logToEvict = slicedLog.pollFirst();
+			logToEvict.clear();
+		}
+
+		public Iterable<StreamRecord<IN>> getReplayLog() throws Exception {
+			final List<String> logSlices = new ArrayList<>(getOperatorStateBackend().getRegisteredStateNames());
+			Collections.sort(logSlices, new Comparator<String>() {
+				@Override
+				public int compare(String o1, String o2) {
+					return Long.valueOf(o1).compareTo(Long.valueOf(o2));
+				}
+			});
+
+			final List<Iterator<StreamRecord<IN>>> wrappedIterators = new ArrayList<>();
+			for (String splitID : logSlices) {
+				wrappedIterators.add(getOperatorStateBackend()
+					.getOperatorState(new ListStateDescriptor<>(splitID,
+						config.<StreamRecord<IN>>getTypeSerializerOut(getUserCodeClassloader()))).get().iterator());
+			}
+
+			if (wrappedIterators.size() == 0) {
+				return new Iterable<StreamRecord<IN>>() {
+					@Override
+					public Iterator<StreamRecord<IN>> iterator() {
+						return Collections.emptyListIterator();
+					}
+				};
+			}
+
+			return new Iterable<StreamRecord<IN>>() {
+				@Override
+				public Iterator<StreamRecord<IN>> iterator() {
+
+					return new Iterator<StreamRecord<IN>>() {
+						int indx = 0;
+						Iterator<StreamRecord<IN>> currentIterator = wrappedIterators.get(0);
+
+						@Override
+						public boolean hasNext() {
+							if (!currentIterator.hasNext()) {
+								progressLog();
+							}
+							return currentIterator.hasNext();
+						}
+
+						@Override
+						public StreamRecord<IN> next() {
+							if (!currentIterator.hasNext() && indx < wrappedIterators.size()) {
+								progressLog();
+							}
+							return currentIterator.next();
+						}
+
+						private void progressLog() {
+							while (!currentIterator.hasNext() && ++indx < wrappedIterators.size()) {
+								currentIterator = wrappedIterators.get(indx);
+							}
+						}
+
+						@Override
+						public void remove() {
+							throw new UnsupportedOperationException();
+						}
+
+					};
+				}
+			};
+		}
+
+		public void clearLog() throws Exception {
+			for (String outputLogs : getOperatorStateBackend().getRegisteredStateNames()) {
```

Comment:
It's kind of bad that we can't remove the state completely and keep iterating over them when replaying the log...